kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/MaybeAsyncProcessor.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2024 Responsive Computing, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package dev.responsive.kafka.api.async.internals;
+
+import static dev.responsive.kafka.api.config.ResponsiveConfig.ASYNC_THREAD_POOL_SIZE_CONFIG;
+import static dev.responsive.kafka.api.config.ResponsiveConfig.responsiveConfig;
+
+import dev.responsive.kafka.api.async.internals.stores.AbstractAsyncStoreBuilder;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.kafka.streams.processor.api.FixedKeyProcessor;
+import org.apache.kafka.streams.processor.api.FixedKeyProcessorContext;
+import org.apache.kafka.streams.processor.api.FixedKeyRecord;
+import org.apache.kafka.streams.processor.api.Processor;
+import org.apache.kafka.streams.processor.api.ProcessorContext;
+import org.apache.kafka.streams.processor.api.Record;
+import org.apache.kafka.streams.processor.internals.InternalProcessorContext;
+
+/**
+ * This class is in charge of two things:
+ * 1. deduplicating and delegating code between the regular and fixed-key processor versions
+ * 2. disabling async processing altogether if the thread pool size is 0
+ */
+public class MaybeAsyncProcessor<KIn, VIn, KOut, VOut>
```

Comment:
not the final name for this -- plan to rename it to "AsyncProcessor" or similar. See [this comment](https://github.com/responsivedev/responsive-pub/pull/306/files#r1630614463)